equinox/nn/_embedding.py

```diff
@@ -89,3 +93,253 @@ def __call__(
                 "`eqx.nn.Embedding()(x)` should be called with a scalar index `x`. "
                 "Use `jax.vmap` if you would like to index with multiple values."
             )
+
+
+class RotaryPositionalEmbedding(Module):
+    embedding_size: int = field(static=True)
+    max_seq_len: Optional[int] = field(static=True)
+    freqs_cis: Optional[Complex[Array, "embedding_size/2"]] = field(static=True)
+
+    def __init__(
+        self,
+        embedding_size: int,
+        max_seq_len: Optional[int] = None,
+        *,
+        key: Optional[PRNGKeyArray] = None,
+        **kwargs,
+    ):
+        """**Arguments:**
+
+        `RotaryPositionalEmbedding` requires:
+
+        - `embedding_size`: Size of the token embeddings. Must be non-negative.
+        - `max_seq_len`: The maximum sequence length. Must be non-negative if provided.
+        - `key`: Not used; provided for compatibility with the rest of the Equinox API.
+            (Keyword only argument.)
+        """
+        super().__init__(**kwargs)
+        if embedding_size < 0:
+            raise ValueError("embedding_size must not be negative.")
+        if max_seq_len is not None and max_seq_len < 0:
+            raise ValueError("max_seq_len must not be negative.")
+        self.embedding_size = embedding_size
+        self.max_seq_len = max_seq_len
+
+        if self.max_seq_len is not None:
+            if (self.embedding_size, self.max_seq_len) in internal_rope_embedding_cache:
+                self.freqs_cis = internal_rope_embedding_cache[
+                    (self.embedding_size, self.max_seq_len)
+                ]
+            else:
+                self.freqs_cis = self.precompute_freqs_cis(
+                    embedding_size, self.max_seq_len
+                )
+                internal_rope_embedding_cache[
+                    (self.embedding_size, self.max_seq_len)
+                ] = self.freqs_cis
+        else:
+            self.freqs_cis = None
+
+    @staticmethod
+    def negate_half(x: Float[Array, "max_seq_len embedding_size"]):
+        d_2 = x.shape[-1] // 2
+        return jnp.concatenate([x[..., :d_2], -x[..., d_2:]], axis=-1)
+
+    @staticmethod
+    def precompute_freqs_cis(
+        embedding_size: int, end: int, theta: float = 10000.0
+    ) -> Complex[Array, "end embedding_size/2"]:
+        def polar(abs, angle):
+            return jnp.array(
+                abs * jnp.cos(angle) + abs * jnp.sin(angle) * 1j, dtype=jnp.complex64
+            )
```

Comment:
I think this function is only used once, so it can be inlined. Moreover I'm guessing the `jnp.array` isn't necessary: probably just `freqs_cis = jnp.cos(freqs_outer) + 1j * jnp.sin(freqs_outer)` will work?